doc/mgmt/gnmi/gNMI_Subscription_for_YangData.md

```diff
@@ -0,0 +1,1183 @@
+
+## 1 Feature Overview
+
+### 1.1 Introduction
+
+SONiC Telemetry service suports gNMI Get, Set and Subscribe RPCs for DB paths and sonic-yang based paths.
+It also suppots Get and Set for OpenConfig and IETF yang based paths that are part of **sonic-mgmt-common** repository.
+This design document describes proposed enhancements to support gNMI Subscribe RPC for such YANG paths.
+
+### 1.2 Requirements
+
+#### 1.2.1 ON_CHANGE subscription for eligible paths
+
+Infrastructure should support ON_CHANGE subscription for YANG paths based on redis keyspace notifications.
+It should provide APIs for apps to specify ON_CHANGE unsupported paths.
+An `INVALID_ARGUMENT` status should be returned if the target path or any of its descendent paths do not support ON_CHANGE.
+
+#### 1.2.2 SAMPLE subscription for all paths
+
+Infrastructure should support SAMPLE subscription for all YANG paths.
+It should provide APIs for apps to indicate minimum `sample_interval` supported for the path.
+An `INVALID_ARGUMENT` status should be returned if the minimum `sample_interval` of target path or any of its descendent paths is more than the requested `sample_interval`.
+
+#### 1.2.3 TARGET_DEFINED subscription for all paths
+
+Infrastructure should support TARGET_DEFINED subscription for all YANG paths.
+Subscribe request should be treated as ON_CHANGE or SAMPLE based on the app's preferences for the target path.
+It should split into multiple requests if the target path supports ON_CHANGE
+but some of its descendent paths do not.
+
+#### 1.2.4 POLL and ONCE subscriptions
+
+Infrastructure should support POLL and ONCE subscriptions for all YANG paths.
+
+#### 1.2.5 Support Wildcard Keys
+
+Infrastructure should support wild cards in **path key** for all subscription modes.
+Example: `/openconfig-interfaces:interfaces/interface[name=*]/config`.
+Path can contain any number of wildcard key values.
+Apps should be allowed to indicate wildcard unsupported paths.
+An `INVALID_ARGUMENT` status should be returned if wildcard key cannot be supported for the target path.
+
+Wildcard in path element (like `/interfaces/*/config` or `/interfaces/.../config`) is a stretch goal.
+
+#### 1.2.6 Scalar encoding for telemetry updates
+
+All telemetry updates should be encoded as gNMI scalar types.
+Each update entry should be a {leaf path, scalar value} pair.
+Scalar type encoding is explained in gNMI specification [section 2.2.3](https://github.com/openconfig/reference/blob/master/rpc/gnmi/gnmi-specification.md#223-node-values).
+
+#### 1.2.7 Co-exist with existing gNMI server functionality
+
+Proposed enhancements should not affect any of the existing functionalities of the SONiC Telemetry Service.
+
+### 1.3 Translib Overview
+
+Translib is a golang library for reading and writing YANG model based data to redis DB or non-DB data source.
+Applications would plugin the YANG models and their translation code into the translib.
```

Comment:
>Applications [](http://example.com/codeflow?start=0&length=12)

The term "Applications" is easy to mislead to user mode process running inside sonic.

Reply:
Will try to rephrase